homeassistant/components/sensor/recorder.py

```diff
@@ -137,6 +156,47 @@ def _time_weighted_average(
     return accumulated / (end - start).total_seconds()
 
 
+def _time_weighted_circular_mean(
+    fstates: list[tuple[float, State]], start: datetime.datetime, end: datetime.datetime
+) -> float:
+    """Calculate a time weighted circular mean.
+
+    The circular mean is calculated by weighting the states by duration in seconds between
+    state changes.
+    Note: there's no interpolation of values between state changes.
+    """
+    old_fstate: float | None = None
+    old_start_time: datetime.datetime | None = None
+    values: list[tuple[float, float]] = []
+
+    for fstate, state in fstates:
+        # The recorder will give us the last known state, which may be well
+        # before the requested start time for the statistics
+        start_time = max(state.last_updated, start)
+        if old_start_time is None:
+            # Adjust start time, if there was no last known state
+            start = start_time
+        else:
+            duration = (start_time - old_start_time).total_seconds()
+            if duration != 0:
+                # No need to add if duration is 0 as it won't affect the result
+                # Append same value for each second between state changes
+                assert old_fstate is not None
+                values.append((old_fstate, duration))
+
+        old_fstate = fstate
+        old_start_time = start_time
+
+    if old_fstate is not None:
+        # Accumulate the value, weighted by duration until end of the period
```

Comment:
There's no accumulation happening here unlike when calculating the arithmetic mean above, please fix the comment